File: backend/src/main/java/info/gamewise/lor/videos/entity/VideoMapper.java
package info.gamewise.lor.videos.entity;

import com.github.brunosc.lor.domain.LoRRegion;
import info.gamewise.lor.videos.domain.LoRVideo;
import info.gamewise.lor.videos.domain.VideoChampion;
import info.gamewise.lor.videos.domain.VideoChannel;
import info.gamewise.lor.videos.domain.VideoRegion;
import info.gamewise.lor.videos.port.out.GetChampionsPort;
import info.gamewise.lor.videos.port.out.GetChannelsPort;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Component;

import java.util.Objects;
import java.util.Set;
import java.util.TreeSet;

import static java.util.stream.Collectors.toUnmodifiableSet;

@Component
class VideoMapper {

    private final GetChannelsPort getChannelsPort;
    private final GetChampionsPort getChampionsPort;

    VideoMapper(GetChannelsPort getChannelsPort, GetChampionsPort getChampionsPort) {
        this.getChannelsPort = getChannelsPort;
        this.getChampionsPort = getChampionsPort;
    }

    Page<LoRVideo> toDomainPage(Page<VideoJpaEntity> page, Pageable pageable) {
        final var videos = page.getContent()
                .stream()
                .map(this::toDomain).toList();

        return new PageImpl<>(videos, pageable, page.getTotalElements());
    }

    LoRVideo toDomain(VideoJpaEntity entity) {
        return new LoRVideo(
                entity.getUrl(),
                entity.getTitle(),
                entity.getDeckCode(),
                mapChannel(entity.getChannel()),
                mapChampions(entity.getChampions()),
                domainRegions(entity.getRegions()),
                entity.getPublishedAt(),
                entity.getThumbnails().getHigh()
        );
    }

    private Set<VideoChampion> mapChampions(Set<String> champions) {
        return champions
                .stream()
                .map(champion -> getChampionsPort.getChampionByName(champion).orElse(null))
                .filter(Objects::nonNull)
                .map(VideoChampion::new)
                .collect(toUnmodifiableSet());
    }

    private Set<VideoRegion> domainRegions(Set<LoRRegion> regions) {
        Set<VideoRegion> domainRegions = regions
                .stream()
                .map(VideoRegion::new)
                .collect(toUnmodifiableSet());

        return new TreeSet<>(domainRegions);
    }

    private VideoChannel mapChannel(String channel) {
        return getChannelsPort.getChannelByCode(channel)
                .map(VideoChannel::new)
                .orElse(null);
    }

}

File: backend/src/main/kotlin/info/gamewise/lor/videos/entity/VideoMapper.kt
package info.gamewise.lor.videos.entity

import com.github.brunosc.lor.domain.LoRRegion
import info.gamewise.lor.videos.domain.LoRVideo
import info.gamewise.lor.videos.domain.VideoChampion
import info.gamewise.lor.videos.domain.VideoChannel
import info.gamewise.lor.videos.domain.VideoRegion
import info.gamewise.lor.videos.port.out.GetChampionsPort
import info.gamewise.lor.videos.port.out.GetChannelsPort
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageImpl
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Component
import java.util.*
import java.util.stream.Collectors

@Component
internal class VideoMapper(
    private val getChannelsPort: GetChannelsPort,
    private val getChampionsPort: GetChampionsPort
) {

    fun toDomainPage(page: Page<VideoJpaEntity>, pageable: Pageable?): Page<LoRVideo> {
        val videos = page.content.map { entity: VideoJpaEntity -> toDomain(entity) }
        return PageImpl(videos, pageable!!, page.totalElements)
    }

    fun toDomain(entity: VideoJpaEntity): LoRVideo {
        return LoRVideo(
            url = entity.url!!,
            title = entity.title!!,
            deckCode = entity.deckCode!!,
            channel = mapChannel(entity.channel!!)!!,
            champions = mapChampions(entity.champions!!),
            regions = domainRegions(entity.regions),
            publishedAt = entity.publishedAt!!,
            thumbnail = entity.thumbnails!!.high
        )
    }

    private fun mapChampions(champions: Set<String>): Set<VideoChampion> {
        return champions
            .mapNotNull { getChampionsPort.getChampionByName(it) }
            .map { VideoChampion(it) }
            .toSet()
    }

    private fun domainRegions(regions: Set<LoRRegion>?): Set<VideoRegion> {
        val domainRegions = regions!!
            .stream()
            .map { region: LoRRegion? -> VideoRegion(region!!) }
            .collect(Collectors.toUnmodifiableSet())
        return TreeSet(domainRegions)
    }

    private fun mapChannel(channelCode: String): VideoChannel? {
        val channel = getChannelsPort.getChannelByCode(channelCode)
        return if (channel != null) VideoChannel(channel) else null
    }
}